### PR TITLE
many: support extended classic models that omit kernel/gadget

### PR DESCRIPTION
**Pull request by pedronis**
redefine IsCoreBoot as the model having a kernel

there are at least tests that show we can seed such systems

the changes ind devicemgr.go prompted fixes for some remodel tests that were
confusingly running wiht OnClassic true though they were using Core models

### CODE REVIEW

**Comment by pedronis** [comment]
based on #12210

**Review by alfonsosanchezbeato — COMMENTED** [review]
Some style remarks. I have a question about this, it looks like with these changes it is possible to have `modeenv`while not having a kernel, what is the use case for this? Seems a bit strange at first sight.

File: overlord/devicestate/devicectx.go [hunk 1]
```diff
@@ -99,16 +99,13 @@ func (dc groundDeviceContext) RunMode() bool {
...
 // HasModeenv is true if the grade is set
-// TODO:UC20: will classic devices with uc20 models have a modeenv? I think so?
 func (dc groundDeviceContext) HasModeenv() bool {
 	return dc.model.Grade() != asserts.ModelGradeUnset
 }
 
-// IsCoreBoot is true when there are modes, or when there are not but
-// we are not in classic (UC16/18 case). If true this implies that a
-// kernel snap is in use.
+// IsCoreBoot is true if the model sports a kernel.
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/devicestate/devicectx.go:106]: no, I meant:

sport verb
1.
wear or display (a distinctive item).

File: asserts/model.go [hunk 2]
```diff
@@ -781,11 +781,20 @@ func assembleModel(assert assertionBase) (Assertion, error) {
 		if err != nil {
 			return nil, err
 		}
-		if modSnaps.gadget == nil {
-			return nil, fmt.Errorf(`one "snaps" header entry must specify the model gadget`)
-		}
-		if modSnaps.kernel == nil {
-			return nil, fmt.Errorf(`one "snaps" header entry must specify the model kernel`)
+		if !classic {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply asserts/model.go:784]: I know, but classic is the "uncommon" case

File: asserts/model.go [hunk 3]
```diff
@@ -781,11 +781,20 @@ func assembleModel(assert assertionBase) (Assertion, error) {
...
+			if modSnaps.gadget == nil {
+				return nil, fmt.Errorf(`one "snaps" header entry must specify the model gadget`)
+			}
+			if modSnaps.kernel == nil {
+				return nil, fmt.Errorf(`one "snaps" header entry must specify the model kernel`)
+			}
+		} else {
+			hasKernel := modSnaps.kernel != nil
+			if modSnaps.gadget == nil {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply asserts/model.go:793]: done for both stanzas

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks for the changes

File: overlord/devicestate/devicectx.go [hunk 4]
```diff
@@ -99,16 +99,13 @@ func (dc groundDeviceContext) RunMode() bool {
...
 func (d *groundDeviceContext) IsCoreBoot() bool {
-	return d.HasModeenv() || !d.Classic()
+	return d.model.Kernel() != ""
```
Comment (mvo5) [inline overlord/devicestate/devicectx.go:108]: This is nice

File: overlord/snapstate/snapstatetest/devicectx.go [hunk 5]
```diff
@@ -75,7 +75,7 @@ func (dc *TrivialDeviceContext) HasModeenv() bool {
 }
 
 func (d *TrivialDeviceContext) IsCoreBoot() bool {
-	return d.HasModeenv() || !d.Classic()
+	return d.Model().Kernel() != ""
```
Comment (mvo5) [inline overlord/snapstate/snapstatetest/devicectx.go:78]: Nice

**Review by mvo5 — APPROVED** [review]
Thanks, this looks fine

**Pull request merged**